pkg/compiler/analysis.go

```diff
@@ -88,43 +65,64 @@ func (c *codegen) traverseGlobals() (int, int, int) {
 	if hasDefer {
 		n++
 	}
-	if n+nConst != 0 || initLocals > -1 {
-		if n > 255 {
-			c.prog.BinWriter.Err = errors.New("too many global variables")
-			return 0, initLocals, deployLocals
-		}
-		if n != 0 {
-			emit.Instruction(c.prog.BinWriter, opcode.INITSSLOT, []byte{byte(n)})
-		}
-		if initLocals > 0 {
-			emit.Instruction(c.prog.BinWriter, opcode.INITSLOT, []byte{byte(initLocals), 0})
-		}
-		seenBefore := false
-		c.ForEachPackage(func(pkg *loader.PackageInfo) {
-			if n+nConst > 0 {
-				for _, f := range pkg.Files {
-					c.fillImportMap(f, pkg.Pkg)
-					c.convertGlobals(f, pkg.Pkg)
-				}
+
+	if n > 255 {
+		c.prog.BinWriter.Err = errors.New("too many global variables")
+		return hasDeploy
+	}
+
+	if n != 0 {
+		emit.Instruction(c.prog.BinWriter, opcode.INITSSLOT, []byte{byte(n)})
+	}
+
+	initOffset := c.prog.Len()
+	emit.Instruction(c.prog.BinWriter, opcode.INITSLOT, []byte{0, 0})
+
+	lastCnt, maxCnt := -1, -1
+	c.ForEachPackage(func(pkg *loader.PackageInfo) {
+		if n+nConst > 0 {
+			for _, f := range pkg.Files {
+				c.fillImportMap(f, pkg.Pkg)
+				c.convertGlobals(f, pkg.Pkg)
 			}
-			if initLocals > -1 {
-				for _, f := range pkg.Files {
-					c.fillImportMap(f, pkg.Pkg)
-					seenBefore = c.convertInitFuncs(f, pkg.Pkg, seenBefore) || seenBefore
-				}
+		}
+		for _, f := range pkg.Files {
+			c.fillImportMap(f, pkg.Pkg)
+
+			var currMax int
+			lastCnt, currMax = c.convertInitFuncs(f, pkg.Pkg, lastCnt)
+			if currMax > maxCnt {
+				maxCnt = currMax
 			}
-			// because we reuse `convertFuncDecl` for init funcs,
-			// we need to cleare scope, so that global variables
-			// encountered after will be recognized as globals.
-			c.scope = nil
-		})
-		// store auxiliary variables after all others.
-		if hasDefer {
-			c.exceptionIndex = len(c.globals)
-			c.globals[exceptionVarName] = c.exceptionIndex
 		}
+		// because we reuse `convertFuncDecl` for init funcs,
+		// we need to cleare scope, so that global variables
+		// encountered after will be recognized as globals.
+		c.scope = nil
+	})
+
+	// This is a small hack to remove complicated logic in `shortenJumps`.
+	// We can decide if initialization function is empty by checking if
+	// some code was emitted during traversal. Note that if a global variable
+	// is defined outside of `init()`, it will be  initialized
+	// (may be with default value) and hence code will be emitted.
+	if initOffset+3 == c.prog.Len() {
+		c.prog.Reset()
```

Comment:
Doesn't this imply `initOffset == 0`? Is it always the case?

Reply:
I thought that all globals are initialized, but we use one slot for `defer` and it _can_ be meaningfully used with no globals (to recover from panic). Fixed this and added a test.